### PR TITLE
Add spec for small Rational passed to Time::at

### PR DESCRIPTION
**Pull request by wildmaples**
Fix: https://github.com/oracle/truffleruby/issues/2286

Changes: 
* Add a spec covering small Rational objects
* Add a fix:
  * Reassigns the `sec` and `usec` variables within the method so it continues to evaluate in the same fashion as `Time::at(seconds, microseconds_with_frac)`.
  * The Rational object is `divmod` into the denominator and numerator. The numerator is the `sec` and the denominator is the fractional seconds which is then converted into `usec` accordingly.

### CODE REVIEW

File: src/main/ruby/truffleruby/core/time.rb [hunk 1]
```diff
@@ -336,7 +336,13 @@ def at(sec, usec=undefined, unit=undefined, **kwargs)
...
         end
 
-      usec = 0 if Primitive.undefined?(usec)
+      if Primitive.undefined?(usec) && Primitive.object_kind_of?(sec, Rational)
+        sec, fractional_seconds = sec.divmod(1)
+        micro_fractional_seconds = fractional_seconds * 1_000_000
+        usec = micro_fractional_seconds.to_i
+      elsif Primitive.undefined?(usec)
+        usec = 0 if Primitive.undefined?(usec)
+      end
```
Comment (tomstuart) [inline src/main/ruby/truffleruby/core/time.rb:345]: This is an improvement for microseconds but still loses precision with nanoseconds:

```
>> t1 = Time.at(Rational(1_486_570_508_539_759, 1_000_000_000))
=> 1970-01-18 05:56:10.508539 +0100
>> t2 = Time.at(1_486_570, 508_539_759, :nsec)
=> 1970-01-18 05:56:10.508539759 +0100
>> t1.usec == t2.usec
=> true
>> t1.nsec == t2.nsec
=> false
>> t1.nsec
=> 508539000
>> t2.nsec
=> 508539759
```

In principle that could be a problem for a separate PR, but it’s broken (at least) [a `Time#-` spec](https://github.com/oracle/truffleruby/runs/2056645766#step:7:34).

File: src/main/ruby/truffleruby/core/time.rb [hunk 2]
```diff
@@ -336,7 +336,13 @@ def at(sec, usec=undefined, unit=undefined, **kwargs)
           raise ArgumentError, "unexpected unit: #{unit}"
         end
 
-      usec = 0 if Primitive.undefined?(usec)
+      if Primitive.undefined?(usec) && Primitive.object_kind_of?(sec, Rational)
+        sec, fractional_seconds = sec.divmod(1)
+        micro_fractional_seconds = fractional_seconds * 1_000_000
+        usec = micro_fractional_seconds.to_i
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/time.rb:342]: This won't work if the Rational has non-zero nanoseconds though, right?
Also, `usec` will be impacted by `second_arg_scale` which seems incorrect.

Would it work to fix the computation below by using `divmod(1)` instead of `to_i` + `% 1.0`?

File: spec/ruby/core/time/at_spec.rb [hunk 3]
```diff
@@ -38,6 +38,16 @@
         Time.at(BigDecimal('1.1')).to_f.should == 1.1
       end
     end
+
+    describe "passed Rational" do
+      it "returns Time with correct nanosecond and microsecond" do
+        t1 = Time.at(Rational(1_486_570_508_539_759, 1_000_000))
+        t2 = Time.at(1_486_570_508, 539_759)
+        
+        t1.usec.should == t2.usec
```
Comment (eregon) [inline spec/ruby/core/time/at_spec.rb:47]: I think we should just test `t1.nsec.should == 539_759_000` here, and we should test nanoseconds too so `t1.nsec.should == 539_759_123`.
Otherwise maybe both ways to build a Time are incorrect in the same way and the spec would pass.

File: src/main/ruby/truffleruby/core/time.rb (hunk 2 above)
Comment (wildmaples) [reply src/main/ruby/truffleruby/core/time.rb]: I updated the fix and added an early return for the Rational case. In the case where `Rational` is used, there is no need for the logic (like `second_arg_scale`) following the if-block. 

This fix is by no means elegant, but I couldn't see a better way to integrate the fix without refactoring the entire method. What do you think?

File: src/main/ruby/truffleruby/core/time.rb (hunk 1 above)
Comment (wildmaples) [reply src/main/ruby/truffleruby/core/time.rb:349]: Fixed in latest version of the commit!

File: src/main/ruby/truffleruby/core/time.rb [hunk 4]
```diff
@@ -336,7 +336,17 @@ def at(sec, usec=undefined, unit=undefined, **kwargs)
           raise ArgumentError, "unexpected unit: #{unit}"
         end
 
-      usec = 0 if Primitive.undefined?(usec)
+      if Primitive.undefined?(usec) && Primitive.object_kind_of?(sec, Rational)
+        rational = Truffle::Type.coerce_to_exact_num(sec)
+        seconds, fractional_seconds = rational.divmod(1)
+        nano_fractional_seconds = fractional_seconds * 1_000_000_000
+
+        time = Primitive.time_at self, seconds.to_i, nano_fractional_seconds.to_i
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/time.rb:344]: `seconds` is already an Integer here, no? (I can do it)

**Review by eregon — APPROVED** [review]
Looks good, and the logic for Rational is indeed much simpler that way.
We should probably refactor this method at some point, but no need now.

**Comment by eregon** [comment]
I tried changing
```ruby
      sec = s.to_i
      nsec_frac = s % 1.0
```
to
```ruby
      sec, nsec_frac = s.divmod(1)
```

and that seems to work just fine for Rational.
Did you try that, was there any issue with it?
It seems to pass the core/time specs at least.

I'll try CI on that (on top of your changes). I also renamed a few variables for clarity.

**Comment by wildmaples** [comment]
> Did you try that, was there any issue with it?

I just tried it locally and had no issues. Thanks, Benoit!
